File: p2p-ws-sdk-ts/src/xor.ts
export function xorNoWrap(plain: Uint8Array, keyfile: Uint8Array, offset: number): Uint8Array {
  if (offset < 0 || offset >= keyfile.length) {
    throw new Error("offset out of range")
  }
  if (offset + plain.length > keyfile.length) {
    throw new Error("offset+plainLen exceeds keyLen (no wrap)")
  }
  const out = new Uint8Array(plain.length)
  for (let i = 0; i < plain.length; i++) {
    out[i] = plain[i] ^ keyfile[offset + i]
  }
  return out
}

File: p2p-ws-sdk-python/src/p2p_ws_sdk/xor.py
from __future__ import annotations


def xor_no_wrap(plain: bytes, keyfile: bytes | bytearray | memoryview, offset: int) -> bytes:
    if offset < 0 or offset >= len(keyfile):
        raise ValueError("offset out of range")
    if offset + len(plain) > len(keyfile):
        raise ValueError("offset+plainLen exceeds keyLen (no wrap)")
    out = bytearray(len(plain))
    for i, b in enumerate(plain):
        out[i] = b ^ keyfile[offset + i]
    return bytes(out)


def xor_repeat(data: bytes, key: bytes) -> bytes:
    if not key:
        raise ValueError("key must not be empty")
    out = bytearray(len(data))
    klen = len(key)
    for i, b in enumerate(data):
        out[i] = b ^ key[i % klen]
    return bytes(out)
